types/meow/index.d.ts

```diff
@@ -26,6 +26,10 @@ declare namespace meow {
         flags?: buildOptions.Options;
         autoHelp?: boolean;
         autoVersion?: boolean;
+        // Caution: explicitly specifying undefined for booleanDefault
+        // has different meaning from omitting key itself.
+        // tslint:disable-next-line no-redundant-undefined
+        booleanDefault?: boolean | null | undefined;
```

Comment:
As you've already noted, in terms of typechecking this will behave exactly the same with or without `| undefined`. To the actual developer using these typings, the subtlety will be totally lost. They won't see anything special about this type in the compilation or the editor.

So I think it's a more helpful approach to:
- bring back the `no-redundant-undefined` linting rule
- remove the `| undefined` section of the type
- turn your comment into a JSDoc comment (`/** */`) so that your warning is visible to the user when they're using this typing in an editing environment.

I can merge once you make these changes :)